micro-rdk/src/esp32/conn/network.rs

```diff
@@ -268,12 +270,31 @@ impl Esp32WifiNetwork {
         &self,
         mut networks: Vec<NetworkSetting>,
     ) -> Result<(), WifiManagerError> {
-        // TODO(RSDK-10184): scan available networks first
-        // attempt to connect only if available
+        log::info!("initializing wifi module");
+        // `set_station_mode` ends by calling `connect`, which will error here
+        let _err = self
+            .set_station_mode(NetworkSetting::new("".to_string(), "".to_string(), 0))
+            .await;
+
+        log::info!("scanning available networks");
+        let available: HashSet<String> = HashSet::from_iter(
+            self.scan_networks()
+                .await
+                .unwrap_or_default()
+                .into_iter()
+                .map(|net| net.0.ssid),
```

Comment:
we could do a filter before mapping to only consider networks above a certain signal strength

Reply:
I think below the [-75,-80] range is usually considered the threshold for a negative experience. But it is tricky: if we have multiple candidate networks available, should we pick the stronger one? But if we can see only one, should we still try to connect to it even if the signal is really bad? If we have a higher priority network, but it has a bad signal, should we still prefer it over a slightly lower priority network that is available with a strong signal? If we have multiple networks with the same priority, should we opt for highest signal first? What does linux do? Is there some sort of known algorithm for deciding this?

Reply:
taken from https://askubuntu.com/a/1348319
> 
    - Scan for potential networks
    - Rank the scanned networks by priority and connect to the highest-priority network
    - Rank the prioritized networks by connection strength and connect to the strongest one
    - Rank the networks by order in the configuration file and connect to the first one

Reply:
made a follow-up ticket [RSDK-10612](https://viam.atlassian.net/browse/RSDK-10612?atlOrigin=eyJpIjoiYmY0ZjVjNjY5ZjgwNGNkY2I1Yzk5ZDM2YzFkMTY2NGYiLCJwIjoiaiJ9)